### PR TITLE
Initial WorkspaceHub change for ML SDK

### PR DESCRIPTION
**Pull request by ZhidaLiu**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_hub/hub.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from azure.ai.ml._schema.workspace.customer_managed_key import CustomerManagedKeySchema
+from azure.ai.ml._schema.workspace.identity import IdentitySchema
+from azure.ai.ml._utils.utils import snake_to_pascal
+from azure.ai.ml.constants._common import PublicNetworkAccess
+from azure.ai.ml._schema.workspace.networking import ManagedNetworkSchema
+
+from azure.ai.ml._schema import ExperimentalField
+
+
+class HubSchema(PathAwareSchema):
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_hub/hub.py:18]: Rename to WorkspaceHubSchema

File: sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py [hunk 2]
```diff
@@ -231,6 +231,8 @@ class AzureMLResourceType:
     """Feature store entity resource type."""
     FEATURE_STORE = "feature_store"
     """Feature store resource type."""
+    HUB_WORKSPACE = "hub"
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/constants/_common.py:234]: Don't call it "HubWorkspace". Rename to "WORKSPACE_HUB"

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_hub/_constants.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+HUB_KIND = "hub"
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_hub/_constants.py:5]: rename to "WORKSPACE_HUB_KIND"

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_hub/_constants.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+LEAN_KIND = "lean"
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_hub/_constants.py:6]: rename to: LEAN_WORKSPACE_KIND

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_hub/hub.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        :param tags: Tags of the feature store.
+        :type tags: dict
+        :param display_name: Display name for the feature store. This is non-unique within the resource group.
+        :type display_name: str
+        :param location: The location to create the feature store in.
+            If not specified, the same location as the resource group will be used.
+        :type location: str
+        :param resource_group: Name of resource group to create the feature store in.
+        :type resource_group: str
+        :param managed_network: workspace's Managed Network configuration
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_hub/hub.py:61]: "workspace's" ==> "Workspace Hub's"

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_hub/hub.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        identity: Optional[IdentityConfiguration] = None,
+        primary_user_assigned_identity: Optional[str] = None,
+        **kwargs,
+    ):
+
+        """WorkspaceHub.
+
+        :param name: Name of the feature store.
+        :type name: str
+        :param description: Description of the feature store.
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_hub/hub.py:50]: Please update these comments. Repace "Feature Store" with "Workspace Hub"

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_load_functions.py [hunk 7]
```diff
@@ -757,6 +758,34 @@ def _load_feature_store_entity(
...
+        the relative locations of files referenced in the parsed yaml.
+        Defaults to the inputted source's directory if it is a file or file path input.
+        Defaults to "./" if the source is a stream input with no name value.
+    :type relative_origin: str
+    :param params_override: Fields to overwrite on top of the yaml file.
+        Format is [{"field1": "value1"}, {"field2": "value2"}]
+    :type params_override: List[Dict]
+    :return: Loaded feature store object.
+    :rtype: _FeatureStore
+    """
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_load_functions.py:785]: same here, please update the comments

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/workspace.py [hunk 8]
```diff
@@ -39,4 +39,9 @@ class WorkspaceSchema(PathAwareSchema):
     )
     identity = NestedField(IdentitySchema)
     primary_user_assigned_identity = fields.Str()
+    storage_accounts = fields.List(fields.Str())
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/workspace.py:42]: why do we need to add these in Workspace object?

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_workspace/workspace.py [hunk 9]
```diff
@@ -265,4 +318,9 @@ def _to_rest_object(self) -> RestWorkspace:
             if self.managed_network
             else None,  # pylint: disable=protected-access
             feature_store_Settings=feature_store_Settings,
+            storage_accounts=self.storage_accounts,
+            container_registries=self.container_registries,
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_workspace/workspace.py:322]: same here.

**Review by seanyao1 — CHANGES_REQUESTED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py [hunk 10]
```diff
@@ -114,6 +115,26 @@ def begin_create(
...
+                x is None
+                for x in [
+                    workspace.storage_account,
+                    workspace.container_registry,
+                    workspace.key_vault,
+                    workspace.public_network_access,
+                    workspace.managed_network,
+                    workspace.customer_managed_key,
+                ]
+            ):
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py:130]: Do all of these have to be None to create a project workspace? I thought PNA can be set for project workspaces.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py [hunk 11]
```diff
@@ -114,6 +115,26 @@ def begin_create(
...
+                msg = "To use workspaceHub, only specify location, application insight, encryption and identity."
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py:131]: The error message is not clear. Please update it to:
To create a project workspace with a workspace hub, please only specify name, description, display_name, location, application insight, encryption and identity."

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py [hunk 12]
```diff
@@ -114,6 +115,26 @@ def begin_create(
 
         workspace.resource_group = resource_group
         template, param, resources_being_deployed = self._populate_arm_paramaters(workspace, **kwargs)
+        # check if create with workspace hub request is valid
+        if workspace._kind == PROJECT_WORKSPACE_KIND and workspace.workspace_hub:
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py:119]: what if it's an empty project workspace without a hub?

**Review by seanyao1 — CHANGES_REQUESTED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py (hunk 10 above)
Comment (ZhidaLiu) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py:130]: According to our code, we override the PNA flag with workspaceHub's PNA during creation, so setting a PNA flag here is  confusing.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py (hunk 12 above)
Comment (ZhidaLiu) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py:119]: right, we don't enforce workspace_hub now.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py (hunk 11 above)
Comment (ZhidaLiu) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations_base.py:131]: I just noticed we use workspaceHub's encryption to override project workspace, I will make sure encryption is None as well.

**Pull request merged**